File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0662_Maximum_Width_of_Binary_Tree.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0662_Maximum_Width_of_Binary_Tree {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int widthOfBinaryTree(TreeNode root) {
        return dfs(root, 0, 1, new ArrayList<Integer>(), new ArrayList<Integer>());
    }

    private int dfs(TreeNode root, int level, int order, List<Integer> start, List<Integer> end) {
        if (root == null) {
            return 0;
        }
        if (start.size() == level) {
            start.add(order);
            end.add(order);
        } else {
            end.set(level, order);
        }
        int cur = end.get(level) - start.get(level) + 1;
        int left = dfs(root.left, level + 1, 2 * order, start, end);
        int right = dfs(root.right, level + 1, 2 * order + 1, start, end);
        return Math.max(cur, Math.max(left, right));
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0662_Maximum_Width_of_Binary_Tree.cc
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  int widthOfBinaryTree(TreeNode* root) {
    vector<pair<unsigned long long, unsigned long long>> vec;
    return (int) dfs(root, 0, 1, vec);
  }

  unsigned long long dfs(TreeNode* root, int level, unsigned long long order, vector<pair<unsigned long long, unsigned long long>>& vec){
    if(!root) {
      return 0;
    }
    if(vec.size() == level) {
      vec.push_back({order, order});
    } else {
      vec[level].second = order;
    }
    unsigned long long cur = vec[level].second - vec[level].first + 1;
    unsigned long long left = dfs(root->left, level + 1, 2 * order, vec);
    unsigned long long right = dfs(root->right, level + 1, 2 * order + 1, vec);
    return max({cur, left, right});
  }
};
